Add diagrams about hub feature choices at each layer

GeorgianaElena opened the pull request:
This PR:
- moves the hub options table into its own section
- adds mermaid diagrams for how hub features are chosen at each level

### NOTE
- The levels and features are documented as per the first iteration of [the features table](https://docs.google.com/spreadsheets/d/1ZRCAMyIBOSfPlfHywEKva7WG4-xNVKm18eAWGo_4KJQ/edit?usp=sharing)
- Once other names for level/features are proposed per https://github.com/2i2c-org/meta/issues/628, we should update it, but this is a starting point for a visualization of our current options

GeorgianaElena commented:
Personally, I'm pretty happy with this state diagram visualization of the hub features.
Check it out
- [here](https://mermaid.live/view#pako:eNqVWFFv2zYQ_iuEnrbB6hIP6RqhKJCka1ogaYKkxYAtg0BLtMVWIjWSqucG-e87kpJFSpTi5qE1ye--Ox6Pd0c9RhnPSZREUmFF3lK8EbiKvy0fGIK_j2T7vlmhOH6DLkre5Fd4R4Rd6sdmGWCfJRFnWUakdGDjecvGq6phVO0uGql4Rb9jRTlzBecQhuItVrjFWwmzA9euRztvrQWWVlAWWJA8QTlZ46ZUPSizoJQzkqAlXWYLdHlxu0CNjDPClMDlcbwaw9WWW3iMtzJu5AKd_XlvpLZEqngZkCgEaVXEzddWC2kEr4mRcUSsscZuz7yR8QEgGDa_rs3ovNd7sC4xS8k32LFEr19nBacZefMm6CZD5-CDzjkApA0Zw3qga5KG5SSnGdiapwyit-a8TBBdoxbBGSqa1bS0dcVeFO29iRhHY-Vu7OwV98v7KbsBwb_RvAtio72dMctw0oGz8ykE2UCk96t2bJYamer4SJcDkif3BgSunHMT_v7lH0N11qgCdqr1etr8eY3UZAxXgUjxsa4WV9PFhyu-4ezGQXOBAm5whS6pgn34MmHkp6oeMutQqAtSEbiyNhR8UWs9BikI8C-cMje8e0BGS214qoFctBloVmJDFaibFBjaEfaM3pRmSMy_oTOalw2abcn2Q58r5GxDFdrPiMnnMl7Vsvc8o7i8Hd0HD_SBSUVVo4l6rDnAilQriFy-BoyuBeDIdwQWg0qt93ltSsTECfn2GO2tQNL98AU6Oo28ppngkq_VNOSS801JZtaNk4eWh2PMpBpeYcrSFZbBYjUv3LR3Nq2xgtx1sFJclnwLuVHLy6Gtwejut_aJ4Eqm9txkQesJo59huREbl-RA2YHdHf5pnLO6bDaVrepmVdLMX7NzZpnAoZRTySsrMa2SPuPLmmR0DZJmYZIzsNpZgxlnu4o3_p5GKX-2WQokf8G5Sm1w6jI-D_DaiIFkoqsNqvGGyDDGNEcaY8PZPY8BlXMjfZtmcLq7gNvLyh1aQ8qS0KPRTaHg9xaLfAGVnguF3n-6vppS3PUnWefBVAfHi64zMQPdp8EM9JU_xHGoGIesCzlvL40q0I0kUQiji49n138gVWCFashp0MAoDkNieseAT4cnAzWCMnQL5zPCmJP56aa2ufdndKPNCB1nqTlSPekd_DmWOrI5W9NNY9OyztaKVNBvQUxKkg2S6pjI2HBPyvWABxpC3mwKxLesTQvQBdV8fJst5XTiH4SNBx8a5ibriT1XmOW6OO5-TNZsk7eedjcRVGNIIA1i1t7jD2zND5F51zDoJM938xq6sLsDh36-u3K6X_Me4WLzazsRQ66NC16R2oufecpzgVlWOKwV1jfJtWm4t27DH02T-RwKbH4GpIubQfW4zjW9n4SvLQDYK_JS7f7ZGcqrEBw4fTa5jlBehg1xwCWxzzetHb2lAu4WFED3loZIE3Sz-gJQdA9oOC903mRfif-6sbsKKR1WyPbZJAjOt4IqMMrOxPuZIN4Wuj04VF87x7QSOYV0SBhelbDj1W7-oTC26gclO_v6-xksxGVpeosQ7ukAf-orMtFx3GeQ9bIivHgLOilcIKbCPe_KnuhUz2upDdOqO3sX0NP7GB-ley5TyNIa8BWV0rbN5luL5uimfLHONs0cphgqsmeRWsEE3ZohFHds3rJ0Neyv14JXKVAnXWQ7SLNoauV792OAqQqNktCgzUu2oDCDIP828BDnAvLgDjxx142h1u4O9J8t_r0pB8t0u55qQ2dEfbuB4PLi1vRP48epY5nTGLdnM2jLN17bb19C6BJqeC2RecCgqY5-qMWj8uwb9r_RIoIHPjRAeZRE5mo9REo_-h-iBH7mWHx9iB6YxsF7gd_vWBYla1xKsoiaOu-_du5nSU7BM9f2e6j5LLqIasz-4hwwSjTEDKPkMfovSk6Oly-WR69eHR0vX56e_Ha8XES7KImPXhzpv-PT05e_n5ycnJ6ewn-vXj4tou-G5vjpf-xV1gE) (where it's zoomable)
- or in the docs at https://2i2c-pilot-hubs--2936.org.readthedocs.build/reference/options/
- or below
```mermaid
stateDiagram-v2
    NewHub --> CloudLayer
    CloudLayer --> HubUserAccessLayer
    HubUserAccessLayer --> CommunityCustomizationsLayer
    CommunityCustomizationsLayer --> DataLayer

    state CloudLayer {
        Cluster --> shared: default
        cluster_one: 2i2c, GCP, us-central1-b
        cluster_two: 2i2c-aws-us, AWS, us-west-2
        cluster_three: 2i2c-uk, GCP, europe-west2
        shared --> cluster_one: default
        shared --> cluster_two
        shared --> cluster_three

        state plan_events <<choice>>
        cluster_one --> plan_events
        cluster_two --> plan_events
        cluster_three --> plan_events

        plan_events --> dedicated_nodepool: if events on hub
        plan_events --> shared_nodepool : default no events

        Cluster --> dedicated
        dedicated --> provider
        provider --> GCP: default
        dedicated --> region
        region --> us_west_2: default
    }
    state HubUserAccessLayer {
        [*] --> Authentication
        Authentication--> Username

        state Authentication {
            [*] --> CILogonOAuthenticator : default
            [*] --> GitHubOAuthenticator
            [*] --> TmpAuthenticator : if ephemeral hub

            state auth <<join>>
            state cilogon_authorization <<join>>
            state github_authorization <<join>>


            CILogonOAuthenticator --> auth: authentication
            CILogonOAuthenticator --> cilogon_authorization: authorization
            GitHubOAuthenticator --> github_authorization: authorization

            auth --> SocialProvider
            auth --> InstitutionalProvider: if member of InCommon Federation

            state options <<join>>
            SocialProvider --> options: options
            options --> Microsoft
            options --> Google
            options --> GitHub

            cilogon_authorization --> domain_based: default
            cilogon_authorization --> username_pattern
            cilogon_authorization --> allowed_users

            github_authorization --> GitHubTeams_membership: default
            github_authorization --> GitHubOrgs_membership
            github_authorization --> allowed_users
        }

        state Username {
            [*] --> public
            public --> email : default
            claim: provider specific claim
            public --> claim
            [*] --> anonymous
        }
    }
    state CommunityCustomizationsLayer {
        [*] --> root_option_one
        [*] --> root_option_two
        root_option_one: Hub pages
        root_option_two: Hub domain

        root_option_one --> option_one
        root_option_one --> option_two : only for straightforward, short HTML

        root_option_two --> community_name.cluster_name.2i2c.cloud
        root_option_two --> community_name.2i2c.cloud
        root_option_two --> other: community must set a CNAME that points to the 2i2c domain

        option_one: Hub Login Page
        option_two: (Optional) Other Hub pages
        login_page_option_one: Basic configuration of template sections
        login_page_option_two: Self configuration through own GitHub repo

        state login_options <<join>>
        option_one --> login_options
        login_options --> login_page_option_one: mandatory
        login_options --> login_page_option_two: optional

        login_page_option_one --> OrganizationInfo
        login_page_option_one --> FundedBy

        login_page_option_two --> RepoURL : default 2i2c-org/default-hub-homepage
        login_page_option_two --> RepoBranch : default master

        OrganizationInfo --> OrgName
        OrganizationInfo --> OrgURL
        OrganizationInfo --> logoURL

        FundedBy --> FunderName
        FundedBy --> FunderURL
    }
    state DataLayer {
        [*] --> data_root_option_one
        [*] --> data_root_option_two
        data_root_option_one: Shared Data Directories
        data_root_option_two: Object Storage Buckets

        state data_root_option_one {
            shared_readwrite: shared-readwrite
            shared_public: shared-public
            [*] --> shared_dir: enabled by default
            [*] --> shared_readwrite: enabled by default
            [*] --> shared_public: optional
            [*] --> allusers: optional
        }

        state data_root_option_two {
            [*] --> Scratch
            [*] --> Persistent

            state buckets <<join>>
            Scratch --> buckets
            Persistent --> buckets

            hub_cloud_persmissions: Cloud Permissions
            buckets --> hub_cloud_persmissions

            public_bucket: Publicly accessible
            from_hub: Buckets accessible from the Hub
            outside_hub: Buckets accessible from outside the Hub
            requestor_pays: Requestor Pays

            hub_cloud_persmissions --> outside_hub
            hub_cloud_persmissions --> from_hub : default
            hub_cloud_persmissions --> requestor_pays : GCP only


            outside_hub --> public_bucket
            gg_membership: Google Groups based membership
            outside_hub --> gg_membership : GCP only
        }
    }
```

There are only a few more documented features left , after which I'll mark the PR as ready for review.

Any feedback until it's marked as ready for review is welcomed.

cc @colliand

damianavila commented:
After a quick check, I just want to say, this is awesome, @GeorgianaElena!!

colliand commented:
I agree! This is excellent work. Thanks @GeorgianaElena. Some next reactions...broken out in atomic comments to facilitate further discussion.

colliand commented:
I craved **hover tips** or **tool tips** or additional information. [This issue](https://github.com/quarto-dev/quarto-cli/discussions/1054) suggest that it may be possible to activate tool tips for our site. Tool tips in Mermaid appear to be supported by Quarto. I am not certain if mermaid.js is fully supported in MyST.

colliand commented:
I imagined what might happen if we had **blue**, **red** and **green** hub types. Click **red** could highlight the decision flow across the tree. A next phase of our work toward "standardization" will be to coalesce the granular features into curated groups of features designed for community archetypes. The mermaid diagram provided a nicely visual base for imagining ways to define products.

colliand commented:
Under `Authentication`, we have `GitHub` appearing under `Social Provider` and separately under `GitHubOAuthenticator`. I believe the OAuth path provides more granular control on access (using Teams and Orgs) whereas the `Social Provider` path is default open to anyone with a GitHub account.

colliand commented:
Under `Social Provider`, there are likely other authentications sources 2i2c should offer. For example, I can imagine nice scenarios where we offer login against [ORCID](https://orcid.org/). 

The scope of auth integrations available under CIlogon is flexible and huge. Is this all easily supported by 2i2c or will we need boundaries here? A text description with links (e.g. this [nice answer from @GeorgianaElena](https://github.com/2i2c-org/meta/issues/631#issuecomment-1653418473)) provided helpful guidance.

colliand commented:
This overview of features does not appear to capture the `Hardware profiles`, `Image Selector` and `Node Share` options 2i2c offers on some of our hubs. 

Related to this, I am not sure how we should capture other giant sets of features that we can offer: 

+ [Jupyter-Server-Proxy](https://jupyter-server-proxy.readthedocs.io/en/latest/).
+ linux desktop in a browser tab
+ Sync Thing
+ gh-scoped-creds

GeorgianaElena commented:
> https://github.com/quarto-dev/quarto-cli/discussions/1054 suggest that it may be possible to activate tool tips for our site.

Nice! Thanks for this great suggestion @colliand! I changed the diagram from a state diagram to a flow chart to be able to hack our way into having tooltips. I added one as a test for the node after CILogonOAuthenticator called `authentication` and one on the `InstitutionalProviders`, if you want to take a look @colliand.

Also, I envision that clicking on nodes in the graph will point to relevant documentation and we'll have tooltips as extra info where needed.

> I imagined what might happen if we had blue, red and green hub types.

That would be super nice indeed. Although we're not there yet, I think it's something to strive towards.

> Under Authentication, we have GitHub appearing under Social Provider and separately under GitHubOAuthenticator. I believe the OAuth path provides more granular control on access (using Teams and Orgs) whereas the Social Provider path is default open to anyone with a GitHub account.

- The GitHubOAuthenticator should be the choice when GitHub Teams/Orgs membership is desired to be checked when authorizing users.
- Loging in using GitHub though the CILogonOAuthenticator is also possible, but in this case we can only manage who should be authorized though an allowed list or even through the domain of the email associated with the GitHub account (though this is not yet fully supported upstream, but rather a 2i2c only feature)

> Under Social Provider, there are likely other authentications sources 2i2c should offer. For example, I can imagine nice scenarios where we offer login against [ORCID](https://orcid.org/).

This should already be possible. We're supporting whatever CILogon itself supports and ORCID is on the list. I added it to the chart as well.

> The scope of auth integrations available under CIlogon is flexible and huge. Is this all easily supported by 2i2c or will we need boundaries here?

2i2c supports everything that CILogon supports with no added complexity I believe.

> Aspects of the DataLayer (e.g. Object Storage Buckets) appear to overlap with CloudLayer. I point this out in case we need to change the top layer descriptions as part of that issue.

Yes, this is something to think about and we should indeed update the diagrams when those will be available. But my main goal for this task was to have something tangible that describes the current offering in a way we can understand it better.
And even though we could argue that there are features that could be into multiple categories or features that depend on other choices and are not represented by arrows, I compromised and wrote on some of the arrow comments like "GCP only" to not loose this extra info, but not complicate the diagram at the same time.

> This overview of features does not appear to capture the Hardware profiles, Image Selector and Node Share options 2i2c offers on some of our hubs.

Yes, I added some more categories today, including the one reffering to ProfileLists

> Related to this, I am not sure how we should capture other giant sets of features that we can offer:
>  [Jupyter-Server-Proxy](https://jupyter-server-proxy.readthedocs.io/en/latest/).
> linux desktop in a browser tab

These two should be in a category about the user image I believe. But sorting that out is also part of this quarter's work and tracked by the other goal. I didn't want to touch that yet, as it's quite complex, but it's on the list.

> Sync Thing

I don't actually know what this is

> gh-scoped-creds

I added this to the "integrations" layer

sgibson91 commented:
> > Sync Thing
> 
> I don't actually know what this is

In a similar way that nbgitpuller allows you to sync git repos, syncthing allows you to sync more arbitrary things such as dropbox folders

- https://github.com/yuvipanda/jupyter-syncthing-proxy
- https://syncthing.net/

GeorgianaElena commented:
Thanks @sgibson91! And are there 2i2c hubs with this feature? I feel like I've been under a rock and never heard of it before 🤯

sgibson91 commented:
> And are there 2i2c hubs with this feature?

I think because it's a "install it in your image and it just works" kind of a feature and doesn't require much setup from the engineering side of things? I'm not sure. Yuvi will know more.

colliand commented:
The `hub options table` is not sorting when column headings are clicked. The `list of running hubs` table is rendered similarly and sorts when column headings are clicked.

sgibson91 commented on docs/reference/options.md:
```diff
@@ -0,0 +1,258 @@
+            SocialProvider --> ORCID
+
+            cilogon_authorization -- default --> domain_based[Domain matching]
+            cilogon_authorization --> username_pattern[Useranem pattern matching]
```
```suggestion
            cilogon_authorization --> username_pattern[Username pattern matching]
```

GeorgianaElena commented:
> The hub options table is not sorting when column headings are clicked. The list of running hubs table is rendered similarly and sorts when column headings are clicked.

Thanks for pointing it out, @colliand! I just checked and only the checkboxes columns cannot be sorted, the other ones  (the hub domains and authenticator work). I will look into either finding a way to sort the checkboxes by adding a rule or something, or maybe just disable the sorting for them altogether.

GeorgianaElena commented:
@colliand, I've just updated the PR to make the table sortable on each column. You can check it out at https://2i2c-pilot-hubs--2936.org.readthedocs.build/reference/options/

GeorgianaElena commented:
@2i2c-org/engineering and @colliand, is there anything that I should be doing extra for this first iteration in order to get this PR closer to be merged?

colliand commented:
I think this should be merged. I will share other suggestions for how to build on this work today.

GeorgianaElena commented:
@sgibson91, do you agree with merging this and iterate on the names/categorizations in later tasks?

sgibson91 commented:
Yep, go for it!
